Add heat capacitor strategy

TomTorger opened the pull request:
Adds a new strategy and some documentation to the repository.

Todo:
Might want to add some warning badges :-)

ottopaulsen commented on docs/examples/example-cascade-temperature-control.md:
```diff
@@ -0,0 +1,64 @@
+> A climate entity
+
+## Instructions:
+> Create an "input_number" entity in Home Assistant and name it "cascade_setpoint"
```
If you use backtick-quotes, you get a better formatting of such keywords.

TomTorger replied:
Started to use it  :-)

ottopaulsen commented on docs/examples/example-cascade-temperature-control.md:
````diff
@@ -0,0 +1,64 @@
+
+
+```json:no-line-numbers
+[{"id":"29fde6c61057362b","type":"tab","label":"PID Controller - cascade","disabled":false,"info":"","env":[]},{"id":"61ae0bee1a382102","type":"server-state-changed","z":"29fde6c61057362b","name":"Setpoint","server":"e2dd69fb.8f70a8","version":3,"exposeToHomeAssistant":false,"haConfig":[{"property":"name","value":""},{"property":"icon","value":""}],"entityidfilter":"input_number.cascade_setpoint" ...
````
The flows looks better pretty printed. Since they are very long, it could be a good idea either to wrap them in a details-container or have them in separate files.

A details container is made like this:

```
::: details Title
content here, including code block
:::
```

TomTorger replied:
Done

ottopaulsen commented on src/strategy-heat-capacitor-functions.js:
```diff
@@ -0,0 +1,181 @@
+const { DateTime } = require("luxon");
+//const decreasing_end_prices = require("../test/data/tibber-decreasing-24h.json");
```
Please remove commented code

TomTorger replied:
Removed

ottopaulsen commented on src/strategy-heat-capacitor-functions.js:
```diff
@@ -0,0 +1,181 @@
+
+function calculate_opportunities(prices, pattern, amount){
+    //creating a price vector with minute granularity
+    var temp_price = Array(prices.length*60).fill(0)
```
Use const in stead of var

TomTorger replied:
Changed

ottopaulsen commented on src/strategy-heat-capacitor-functions.js:
```diff
@@ -0,0 +1,181 @@
+function calculate_opportunities(prices, pattern, amount){
+    //creating a price vector with minute granularity
+    var temp_price = Array(prices.length*60).fill(0)
+    var end_index=[];
```
Remove unused variables

TomTorger replied:
Done

ottopaulsen commented on src/strategy-heat-capacitor-functions.js:
```diff
@@ -0,0 +1,181 @@
+    //creating a price vector with minute granularity
+    var temp_price = Array(prices.length*60).fill(0)
+    var end_index=[];
+    for(i=0;i<prices.length;i++){
```
Always use `let` when defining for loop variables. `let i = 0`

TomTorger replied:
Done. At least I think so :-)

ottopaulsen commented on src/strategy-heat-capacitor-functions.js:
```diff
@@ -0,0 +1,181 @@
+    }
+    
+    //Calculate weighted pattern
+    var weight = amount/pattern.reduce((a, b) => a + b, 0) //last calculates the sum of all numbers in the pattern
```
const

TomTorger replied:
Done

ottopaulsen commented on src/strategy-heat-capacitor-functions.js:
```diff
@@ -0,0 +1,181 @@
+    
+    //Calculate weighted pattern
+    var weight = amount/pattern.reduce((a, b) => a + b, 0) //last calculates the sum of all numbers in the pattern
+    var weighted_pattern = pattern.map((x)=>x*weight)
```
const

TomTorger replied:
Done

ottopaulsen commented on src/strategy-heat-capacitor-functions.js:
```diff
@@ -0,0 +1,181 @@
+    
+    //Calculating procurement opportunities. Sliding the pattern over the price vector to find the price for procuring
+    //at time t 
+    dot = (a, b) => a.map((x, i) => a[i] * b[i]).reduce((m, n) => m + n);
```
use const when defining constnts like this

TomTorger replied:
Done

ottopaulsen commented on src/strategy-heat-capacitor-functions.js:
```diff
@@ -0,0 +1,181 @@
+    //Calculating procurement opportunities. Sliding the pattern over the price vector to find the price for procuring
+    //at time t 
+    dot = (a, b) => a.map((x, i) => a[i] * b[i]).reduce((m, n) => m + n);
+    var proc_opp = Array(prices.length*60-pattern.length+1)
```
const

TomTorger replied:
Done

ottopaulsen commented on src/strategy-heat-capacitor-functions.js:
```diff
@@ -0,0 +1,181 @@
+}
+
+
+function test(){ //Set up price data
+    price_data = decreasing_end_prices.priceData //decreasing_end_prices.priceData;
+    //User input
+    time_heat_1c = 55
+    time_cool_1c = 40
+    max_temp_adjustment = 0.61
+    min_saving_NOK_kWh = 0.1
+    debugger
+    let schedule = run_buy_sell_algorithm(price_data, time_heat_1c, time_cool_1c,max_temp_adjustment,min_saving_NOK_kWh)
+    dT = find_temp(DateTime.local(2022, 1, 2),schedule)
+    debugger
+}
+
+//var value = test()
```
This code is not used, so please remove it

TomTorger replied:
Done

ottopaulsen commented on docs/nodes/ps-strategy-heat-capacitor.md:
```diff
@@ -0,0 +1,42 @@
+# ps-strategy-heat-capacitor
+
+<!--
+[ps-strategy-lowest-price](../images/node-ps-strategy-lowest-price.png)
+-->
```
Suppose it shall not be commented out

TomTorger replied:
Done

ottopaulsen commented on docs/nodes/ps-strategy-heat-capacitor.md:
```diff
@@ -0,0 +1,42 @@
+
+The node consumes price information and outputs $\Delta T$ on its first output and the planned schedule and benefit calculations on the secound output. The $\Delta T$ is used to adjust the set-point of a climate entity.
+
+## The impact of **Time +1C**
+This time is used to optimize the timing of when to turn up the heat. An increase in temperature from 22 to 23C will cause an increased electricity consumption for quite some time after the change has occured. For heat-pumps, the air is heated first, then walls, furniture etc., creating a high electricity demand in the first hour or so before the demand slowly reduces down to normal levels. The a ...
+
+To get started, 90 minutes can be used for air heaters. Later, one can study the temperature curves or, if available, the power consumption curves of the climate entity, to achieve a better accuracy.
+
+## Max temperature adjustment
+A value of 0.65 will change the setpoint temperature by $\pm0.65$. Please note that a larger number will indicate a longer heating time (Time +1C = 60 and Max Temp Adj.= 0.75 results in a heating time of $60\times0.75\times 2= 90$ minutes
+
+## Min Savings
```
Use heading level 3 on these headings, since they are part of the config

TomTorger replied:
Changed

ottopaulsen reviewed (changes requested):
Ok, I have reviewed a half code file, and I have some general things I would like you to fix.

1. Du not use `var`. Always use `const` for constants, and `let` for variables that are being changed. Use `let` only when necessary.
2. Always explicitly define constants and variables with const or let. Your IDE should help you with this.
3. Remove unused and commented code.
4. Please install a Prettier-plugin and use Prettier to help you format the code. It is very useful to let the plugin do this for you when saving.

I will review more when you have fixed this, but I have very little time this week.

TomTorger commented on docs/examples/example-cascade-temperature-control.md:
```diff
@@ -0,0 +1,64 @@
+ A cascade temperature controller is a controller which utelizes the input/setpoints of other controllers/climate entities to regulate to its own setpoint
+
+---
+## Requirements:
```
Should be fixed now

TomTorger commented:
Made most corrections i believe :-) 

Will have to take a closer look at the .json flows.

ottopaulsen commented on src/strategy-heat-capacitor.js:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    this.on("input", function (msg) {
+      // Using msg.* to change specific TM property.
+      if (msg.hasOwnProperty("time_heat_1c")) node.time_heat_1c = Number(msg.time_heat_1c);
+      if (msg.hasOwnProperty("time_cool_1c")) node.time_cool_1c = Number(msg.time_cool_1c);
+      if (msg.hasOwnProperty("max_temp_adjustment")) node.max_temp_adjustment = Number(msg.max_temp_adjustment);
+      if (msg.hasOwnProperty("min_saving_NOK_kWh")) node.min_saving_NOK_kWh = Number(msg.min_saving_NOK_kWh);
+
+      if (msg.topic == "configure") {
+        node.setpoint = Number(msg.payload);
+      } else {
```
If I understand this right, it is about sending config as input.
To avoid confusing the users, you should do this the same way as it is done in the other nodes. That is sending the config in a config-object in the payload.

I believe the setpoint also could be sent as config. How about having the setpoint in the config dialog, to set a default value? Or maybe that does not make sense? (I do not fully understand this node, yet :-)

Config-variables sent in the input should definitely have camelCasing as in other nodes, since they are visible to the user.

You should be able to use the `getEffectiveConfig` function in `utils.js` to read the config. It will give you the config either from the message or from the node. It requires that you save the config to the nodes context when the node is loading. See `strategy-lowest-price.js` lines 9-20 how this is done, and how it is used in `handle-input.js` line 6.

Some people are sending config and priceData in the same message, so that should work here too.

TomTorger replied:
> I believe the setpoint also could be sent as config. How about having the setpoint in the config dialog, to set a default value? Or maybe that does not make sense? (I do not fully understand this node, yet :-)

Good point. It would simplify the flow... I've tried to think of a situation where one would like to keep the current design, however, didn't find any very good arguments. Now... Where to put it.. I am tempted to use msg.topic = "config", however, in our current setup it should be msg.config.setpoint... Seems simpler to use .topic, but we should probably follow convention?

![image](https://user-images.githubusercontent.com/49571281/152611752-8f9416df-5d87-4908-a932-5c797f0ebb6d.png)

ottopaulsen replied:
Yes, please follow the save convention as already used. Important that users see consistency. Topic is ment for message topics, like used in MQTT, and is also suited for entity_id as in home assistant. Data could have been used, but I think it is easier when everything you need is in the payload.

But convention is most important anyway.

TomTorger replied:
Done

ottopaulsen commented on test/strategy-heat-capacitor.test.js:
```diff
@@ -0,0 +1,81 @@
+
+describe("Test heat capacitor strategy functions", () => {
+  before(function() {
+    prices =  converted_prices.priceData.slice(0, 1).map((p) => p.value);
+    decreasing_24h_prices =  decreasing_end_prices.priceData.slice(0, 1).map((p) => p.value);
+    start_date =  DateTime.fromISO(converted_prices.priceData[0].start);
+    buy_pattern= Array(Math.round(time_heat_1c*max_temp_adjustment*2)).fill(1)
+    sell_pattern= Array(Math.round(time_cool_1c*max_temp_adjustment*2)).fill(1)
```
Please define those variables before you use them.
You may insert `"use strict";` as the first line in the file to help you find identifiers that are not declared.

ottopaulsen commented on docs/nodes/ps-strategy-heat-capacitor.md:
```diff
@@ -0,0 +1,45 @@
+# ps-strategy-heat-capacitor 
+
```
In the documentation, please describe the following in a similar way as other nodes:

* Input
* Output
* Dynamic config

You should give a better description of how the output can be used.
One should not have to import the examples to understand how it is used.

TomTorger replied:
Todo

TomTorger replied:
Da var det gjort :-)

ottopaulsen commented on docs/examples/example-cascade-temperature-control.md:
````diff
@@ -0,0 +1,77 @@
+<details>
+  <summary>Click to expand!</summary>
+
+```json:no-line-numbers
+[{"id":"29fde6c61057362b","type":"tab","label":"PID Controller - cascade","disabled":false,"info":"","env":[]},{"id":"61ae0bee1a382102","type":"server-state-changed","z":"29fde6c61057362b","name":"Setpoint","server":"e2dd69fb.8f70a8","version":3,"exposeToHomeAssistant":false,"haConfig":[{"property":"name","value":""},{"property":"icon","value":""}],"entityidfilter":"input_number.cascade_setpoint" ...
````
When you export the flow from Node-RED, there is an option to format it before you copy it. See the button down right in the export dialog.

TomTorger replied:
Done

ottopaulsen reviewed (changes requested):
Ok, then I have some more comments.

I would prefer that you used camelCase for identifiers (constant, variables, functions),
as is normal in all modern JavasScript style guides. Not snake_case. That may be a bit of work to change, though :-)

TomTorger commented on src/strategy-heat-capacitor.html:
```diff
@@ -0,0 +1,61 @@
+</script>
+
+<script type="text/html" data-help-name="TM">
+  <p>A node for the reduction of electricity costs by moving heating efforts towards times when electricity is cheap</p>
+
+  <p>.</p>
+
+  <ul>
+    <li>
+      <b>Time +1C [minutes]</b> - Time required by the system to increase 1C <b>msg.topic</b> set to
+      <code>setpoint</code> and <b>msg.payload</b> set to the required floating point value, or by passing the value in
+      <b>msg.setpoint</b>.
+    </li>
+    <li><b>Time -1C [minutes]</b> - Time to loose 1C when no heat is applied <b>msg.prop_band</b>.</li>
+    <li><b>Max temp adj.</b> - Max temperature adjustment</li>
+    <li><b>Min Savings [NOK/kWh]</b> - Min savings per cycle</li>
+  </ul>
+</script>
```
Todo

TomTorger commented:
Still some changes to perform:
-As you suggested, I think I will add the setpoint as an input. It removes the need for a couple of function nodes.
-Some node documentation left.
-Converge on the statistics payload... The second output yields a lot of statistics, however, it is not very well structured. I should hava a look at it.
![image](https://user-images.githubusercontent.com/49571281/152615135-84a60525-0bf9-45bc-abf8-281435ad50d6.png)

ottopaulsen commented:
> Still some changes to perform: -As you suggested, I think I will add the setpoint as an input. It removes the need for a couple of function nodes. -Some node documentation left. -Converge on the statistics payload... The second output yields a lot of statistics, however, it is not very well structured. I should hava a look at it. ![image](https://user-images.githubusercontent.com/49571281/152615135-84a60525-0bf9-45bc-abf8-281435ad50d6.png)

Also, please use the same kind of casing in your output as is used in other nodes.

TomTorger commented on src/strategy-heat-capacitor.js:
```diff
@@ -0,0 +1,50 @@
+const { DateTime } = require("luxon");
+//const { loadDayData } = require("./utils");
```
Done

TomTorger commented on src/strategy-heat-capacitor.js:
```diff
@@ -0,0 +1,50 @@
+//const { loadDayData } = require("./utils");
+const { run_buy_sell_algorithm, find_temp } = require("./strategy-heat-capacitor-functions");
+//const { difference } = require("lodash");
+
```
Done

TomTorger commented on src/strategy-heat-capacitor.html:
```diff
@@ -0,0 +1,61 @@
+  </div>
+</script>
+
+<script type="text/html" data-help-name="TM">
```
Changed

TomTorger commented:
Updated. There are quite a few changes... I've added node tests, updated the documentation, added setpoint support, and boosted temperature functionality. I've also switched to camelcase in the code, however, there are some remaining snake_case in the tests. Will try to fix them in the next round.

ottopaulsen commented:
This is really getting good!

I collect all my comments here now.

Code:

Very good improvements to the code!

I just found a few details:

You should use the roundPrice() function from utils on prices before they are sent to the output. That will round them nicely to 4 decimals. Examples should be updated with 4 decimals. That looks better, and some of the roundings are actually wrong.

In the `strategy-heat-capacitor.test.js` file you are using the Array .at() function. This requires node version 16.6 or later.
Many users are on earlier versions of node, so I recommend not using it yet. All other code runs on node 14, and I sometimes get requests from people using even earlier versions.

Output 3:

The variables that are config should be put in a config object, same as on other nodes.

On the trades, you have sellPrice and buyPrice. What if a sell/buy period lasts more than an hour, will you not get multiple prices then? Or do I misunderstand?

Other nodes have "time" and "version" in the output. These can be useful for debugging. I suggest you do the same. Time is the time the plan is running.

Could you have a schedule too in the output, showing planned time and value that will be sent to output 1? That would make it easier for users to understand what is happening, and some may use it in dashboard etc.

I also think it would be useful to have an hours-array in the output showing start and price. It is very helpful when people are asking for support. Also the "source" attribute should be there to show where prices came from.

Other: 

If you use a spellchecker in vscode, you will see some names getting a warning, like Datetime and Dictlist. I like having as few as possible warnings in the editor.

You have not run prettier on all code. I suggest running it on save.


Doc:

Very good updates to the doc too :-)

But still some things that can be improved:

Please show json example for output 2. There are still 3 outputs, yes?

Could you give a brief explanation of the trades and temperatures arrays? Just so people get what it is.

I think buyIndex and sellIndex also need explanation. 

If you run `npm run docs:dev` you can serve the doc locally and see exactly how it is rendered. I use Vuepress 2 for the docs. There are some very useful custom containers that you could use: https://v2.vuepress.vuejs.org/reference/default-theme/markdown.html#custom-containers

For example, you could use the details custom container instead of the details html tag to get a better styling.

In the examples you are using the blockquote notation. These are not quotes, so maybe some other means could be better. Some places you could just use normal bullet points, other places maybe plain text is better. Run locally to see how it renders. Buy the way, I did not know about the blockquote in Markdown, so thank you for teaching me :-)


Not sure I got everything here. I managed to write all the feedback for then loosing it by a mistake, so I had to write everything again. I may have forgotten something...

TomTorger commented:
Summary:

> You should use the roundPrice() function from utils on prices before they are sent to the output. That will round them nicely to 4 decimals. Examples should be updated with 4 decimals. That looks better, and some of the roundings are actually wrong.`
Done

> In the strategy-heat-capacitor.test.js file you are using the Array .at() function. This requires node version 16.6 or later.
Many users are on earlier versions of node, so I recommend not using it yet. All other code runs on node 14, and I sometimes get requests from people using even earlier versions.

Fixed

>Output 3: 
>The variables that are config should be put in a config object, same as on other nodes.
Done
>On the trades, you have sellPrice and buyPrice. What if a sell/buy period lasts more than an hour, will you not get multiple prices then? Or do I misunderstand?
The sellPrice/buyPrice indicates the average cost pr. kWh. Given that the prices at 03:00 to 04:00 and 04:00 to 05:00 are 1 and 2 NOK/kWh, respectively, a 60 min heating period starting at 03:30 and ending at 04:30, will attain an average price of 1,50 NOK/kWh. The algorithm can also support non-flat procurement patterns, but for now, each minute is given even weight in the sellPrice/buyPrice.

>Other nodes have "time" and "version" in the output. These can be useful for debugging. I suggest you do the same. Time is the time the plan is running.
Done

> Could you have a schedule too in the output, showing planned time and value that will be sent to output 1? That would make it easier for users to understand what is happening, and some may use it in dashboard etc.
I could do that, but the output would be huge... :-) I store potentially 72 hours of priceData, generating potentially 4320 values.

>I also think it would be useful to have an hours-array in the output showing start and price. It is very helpful when people are asking for support. Also the "source" attribute should be there to show where prices came from.
Done

>Other:
>If you use a spellchecker in vscode, you will see some names getting a warning, like Datetime and Dictlist. I like having as few as possible warnings in the editor.
Done

>You have not run prettier on all code. I suggest running it on save.
Done, might have missed some spots...

>Doc:
>Very good updates to the doc too :-)
>But still some things that can be improved:
>Please show json example for output 2. There are still 3 outputs, yes?
Done

>Could you give a brief explanation of the trades and temperatures arrays? Just so people get what it is.
Done

>I think buyIndex and sellIndex also need explanation.
Done... Not too sure of the quality of the explanations :-)

>If you run npm run docs:dev you can serve the doc locally and see exactly how it is rendered. I use Vuepress 2 for the docs. >There are some very useful custom containers that you could use: https://v2.vuepress.vuejs.org/reference/default-theme/markdown.html#custom-containers
>For example, you could use the details custom container instead of the details html tag to get a better styling.
Downloaded and tested. Containers look great, just hope they show up on Github :-)

>In the examples you are using the blockquote notation. These are not quotes, so maybe some other means could be better. >Some places you could just use normal bullet points, other places maybe plain text is better. Run locally to see how it renders. Buy the way, I did not know about the blockquote in Markdown, so thank you for teaching me :-)
Todo: Will give it a go and se if I find a nice way to do this. Not too versed in MD.

> Please add links to the examples in the examples/README.md file
Done

>Not sure I got everything here. I managed to write all the feedback for then loosing it by a mistake, so I had to write everything again. I may have forgotten something...
Hehe... I used Ctrl-C regularly as a back-up while writing this. :-) Feel free to comment if you find/remember something more.

ottopaulsen commented:
Very good work!

I have been trying to get the TeX rendered in Vuepress, but currently without success. Not yet found a plugin for Vuepress 2. I am sure it is possible. Do you have any idea?

About outputting a "schedule", I agree with you that we should not output every minute. However, the data is not changing every minute, maybe it could be possible to output a schedule with only the entries when it is actually changing? We don't need to do it immediately.

Anyway, do you feel the code is ready to be merged, or is there anything else you like to do?

TomTorger commented:
>I have been trying to get the TeX rendered in Vuepress, but currently without success. Not yet found a plugin for Vuepress 2. I am sure it is possible. Do you have any idea?

Hmm... I see that it is not rendering correct on Github either. I will see if I can figure it out/change to another approach.

>About outputting a "schedule", I agree with you that we should not output every minute. However, the data is not changing every minute, maybe it could be possible to output a schedule with only the entries when it is actually changing? We don't need to do it immediately.

The "trades" data is intended for this. It shows the times at which the system moves from maintaining the current temperature, to either reducing or increasing it.

>Anyway, do you feel the code is ready to be merged, or is there anything else you like to do?

Hehe... I would say that it is ready for beta testing? :-) If someone could try to implement it on their system and then give feedback on the documentation and functionality?
